Optimize redundant fail checks

Mingun opened the pull request:
Port of my optimization pegjs/pegjs#400, that reduce many dumb comparisons (that is not cut by minifiers). You can see some of them on the `parser.js` (but enable whitespace-ignored mode on github)

I calculate possible values of the match result and do not generate conditions, that are always `true` or `false` and also didn't generate not used constants.

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -225,7 +225,10 @@ function generateBytecode(ast) {
-  function buildCondition(condCode, thenCode, elseCode) {
+  function buildCondition(match, condCode, thenCode, elseCode) {
+    if (match > 0) { return thenCode; }
+    if (match < 0) { return elseCode; }
```
it's not clear what this does, why the API is replaced, what these arguments mean, why "match" being above or below zero does these things, et cetera.

StoneCypher replied:
"it's an internal api" isn't a good reason for unexplained changes with obcure names that take arithmetic approaches to non-arithmetic problems

this is ioccc code

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -244,6 +247,8 @@ function generateBytecode(ast) {
   }
 
   function buildSimplePredicate(expression, negative, context) {
+    const match = expression.match | 0;
```
what is `expression.match`?  where does it come from?  what does it do?  why is it named this, when none of the things in the language names "match" are a number?  what's with the bitwise or of zero?

StoneCypher replied:
> that guaranties, that the result is integer with value 0, and not undefined if that property doesn't exist

this is absurdly over-tricky code for the desired result.  you are attempting to be an optimizing compiler, and expecting javascript programmers to know your intent, these consequences, and incorrectly believing that transcompilers won't replace this with things without these guarantees.

the desired approach is

```typescript
if (somevariable === 0)
```

This will ensure that the result is a number (javascript doesn't have integers) with value `0`, and not `undefined`

It's also faster, conceptually describes correctly what it does, and doesn't require a trivia course to infer intent

StoneCypher replied:
> This is semi-standard way to say the JS engine that there expected an integer. 

No it isn't.

Stop resolving things without the commenter's permission.  This isn't resolved.

&nbsp;

> Even a whole library grew out of this -- ASM.js

This is not asm.js, this is javascript.  This is why I said "you are attempting to be an optimizing compiler."

I already pointed out a bug this causes.  You continued to explain, instead of understanding the bug.

StoneCypher replied:
>  undefined (== absence of the property) should be treated like the value 0. Your simple check fails on that

This is not correct.

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -254,6 +259,7 @@ function generateBytecode(ast) {
       }),
       [op.SILENT_FAILS_OFF],
       buildCondition(
+        negative ? -match : match,
```
really?

StoneCypher replied:
> When you learn about what the match means (and I add an explanation to this pass), your question will gone, I guess. 

You're the guy who made me explain how typescript worked.  Now you won't explain your own code 😂

Mingun replied:
I do not see what lead you to that assumption in my phrase, which says the exact opposite. If my language not so good (I am ready to admit it, I'm not a native speaker), let me known in a straightforward way, and I'll try to use other words

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -269,13 +275,14 @@ function generateBytecode(ast) {
       [op.UPDATE_SAVED_POS],
       buildCall(functionIndex, 0, context.env, context.sp),
       buildCondition(
+        node.match | 0,
```
why are you bitwise oring node.match with zero?  what does this mean?

StoneCypher replied:
Not really, no.

StoneCypher replied:
It turns out this is a substitute for `(if node.?match === 0)`

This is bad code and should not land.  We are not creating a trivia contest.

Mingun replied:
Not really. `node.?match === 0` produces `true` or `false`, `node.match | 0` produces an integer, which represents ternary logic `<0`, `==0`, and `>0`.

StoneCypher replied:
> Not really. node.?match === 0 produces true or false, node.match | 0 produces an integer, which represents ternary logic <0, ==0, and >0.

This is the kind of answer that would get the PR banned in almost any project

StoneCypher replied:
It is clear that you don't even understand what other people think is the problem here

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -313,7 +320,8 @@ function generateBytecode(ast) {
 
     named(node, context) {
-      const nameIndex = addConst(
+      const match = node.match | 0;
```
why are you bitwise oring node.match with zero? what does this mean?

StoneCypher replied:
Not really, no.

StoneCypher replied:
It turns out this is a substitute for `(if node.?match === 0)`

This is bad code and should not land.  We are not creating a trivia contest.

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -313,7 +320,8 @@ function generateBytecode(ast) {
     named(node, context) {
-      const nameIndex = addConst(
+      const match = node.match | 0;
+      const nameIndex = match > 0 ? null : addConst(
```
what does this mean?

StoneCypher replied:
> This is one-liner that doesn't create a constant if it is not used. As you can see, nameIndex used only in the else part of condition, and when match > 0 that means that condition is always true, so the else part is not generated and the constant is not required.

I can't even follow this "explanation."

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -362,13 +372,17 @@ function generateBytecode(ast) {
       });
-      const functionIndex = addFunctionConst(Object.keys(env), node.code);
+      const match = node.match | 0;
+      const functionIndex = emitCall && match >= 0
```
what does any of this mean?

StoneCypher replied:
Imagine thinking that code where you have to look at bytecode output to understand what it does is acceptable, but calling other people's code low quality and sloppy.

This approach is unreadable and should not be accepted.

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -339,6 +347,8 @@ function generateBytecode(ast) {
           alternatives.length > 1
             ? buildCondition(
+                // If alternative always match, no need to generate code for the next alternatives
+                -(alternatives[0].match | 0),
```
what is the negative or of the first element of alternatives, sub index match, bitwise or of zero?

could you please write code that isn't cryptic?

StoneCypher replied:
> I think, the comment above explaining enough if you know what for match are.

If people are having to ask you what it means, then this is not correct.

Why are you implementing the same mystery logic in a dozen places?  What happens when someone who isn't you has to maintain this?

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -395,6 +408,7 @@ function generateBytecode(ast) {
               action: null
             }),
             buildCondition(
+              elements[0].match | 0,
```
what does this mean?

StoneCypher replied:
No, it wasn't.  You gave a Zorba the Mystic "when you look deep enough into the soul of `match` you will understand."

StoneCypher replied:
It turns out this is `(elements || [])[0]?.match === 0`

Admittedly, getting this idiomatically correct in a way that's clear to a reader and doesn't rely on trivia knowledge of operator behavior isn't so easy.

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -479,6 +493,7 @@ function generateBytecode(ast) {
           action: null
         }),
         buildCondition(
+          node.match | 0,
```
what does this mean?

StoneCypher replied:
No, it wasn't.  You gave a Zorba the Mystic "when you look deep enough into the soul of `match` you will understand."

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -502,6 +517,10 @@ function generateBytecode(ast) {
+          // Check expression match, not the node match
+          // If expression always match, no need to replace FAILED to NULL,
+          // because FAILED will never appeared
+          -(node.expression.match | 0),
```
why is this random code repeated all over the place?

why isn't it behind a method with a clear name that explains what it does?

do you believe this comment tells a reader what you're doing?

Mingun replied:
Explained above.

StoneCypher replied:
No, it wasn't.  You gave a Zorba the Mystic "when you look deep enough into the soul of `match` you will understand."

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -535,6 +554,8 @@ function generateBytecode(ast) {
         expressionCode,
         buildCondition(
+          // Check expression match, not the node match
+          node.expression.match | 0,
```
why is this random code repeated all over the place?

why isn't it behind a method with a clear name that explains what it does?

Mingun replied:
Explained above.

StoneCypher replied:
No, it wasn't.  You gave a Zorba the Mystic "when you look deep enough into the soul of `match` you will understand."

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -551,11 +572,11 @@ function generateBytecode(ast) {
 
     semantic_and(node, context) {
-      return buildSemanticPredicate(node.code, false, context);
+      return buildSemanticPredicate(node, false, context);
```
"buildSemanticPredicate" could mean ***literally anything***

Mingun replied:
Well, that name already used and I'm not mix PR changes with random changes not related to the main goal of my PRs

StoneCypher replied:
okay, that's fair.  withdrawn

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -564,13 +585,15 @@ function generateBytecode(ast) {
-        const stringIndex = addConst("\""
+        const match = node.match | 0;
+        const needConst = match === 0 || (match > 0 && !node.ignoreCase);
+        const stringIndex = needConst ? addConst("\""
```
are you really attempting to manually quote string escaping?

Mingun replied:
Again, that PR not tries to change everything in one commit. Escaping problem is solving in the #117

StoneCypher replied:
It turns out the first line is a substitute for `(if node.?match === 0)`

It's still not clear why match is being controlled with arithmetic, despite asking several times, because I got some monk on the mountain "when you truly understand what match is" to a direct question "why are you doing this"

The third one can easily incur side effects during maintenance because of the over-tricky use of ternary select, where the correct form is actually `if` 

This should not land.  We are not creating a trivia contest.

Mingun replied:
> ... to a direct question "why are you doing this"

As I say, introduction comment coming soon

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -564,13 +585,15 @@ function generateBytecode(ast) {
-        );
-        const expectedIndex = addConst(
+        ) : null;
+        const expectedIndex = match > 0 ? null : addConst(
```
why is match being compared to zero when it's also being bitwise manipulated everywhere?

Mingun replied:
If you look more carefully, you will see that bitwise operators is used to get the value of the `match` variable. This doesn't magically calculate any comparing results for us

StoneCypher replied:
> If you look more carefully

People shouldn't have to look more carefully to understand simple code.  This is an excuse.

&nbsp;

> you will see that bitwise operators is used to get the value of the match variable.

This doesn't make any sense.  Bitwise operators are for performing bitwise operations, not getting values of variables.

Mingun replied:
> People shouldn't have to look more carefully to understand simple code. This is an excuse.

All that their need to look, it is to read an introduction comment that coming soon

StoneCypher commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -614,8 +638,9 @@ function generateBytecode(ast) {
         + "]";
-      const regexpIndex = addConst(regexp);
-      const expectedIndex = addConst(
+      const match = node.match | 0;
```
please don't write code that other people can't read, then repeat it dozens of places across the codebase

Mingun replied:
Explained above.

StoneCypher replied:
No, it wasn't.

StoneCypher replied:
It turns out this is a substitute for `(if node.?match === 0)`

StoneCypher commented on lib/compiler/passes/inference-match-result.js:
```diff
@@ -0,0 +1,108 @@
+  });
+
+  inference(ast);
+}
```
are you attempting to add a nearly 100 line function to the codebase, after telling someone else that their 35 line PR was too complicated for you to follow, and couldn't go in?

Mingun replied:
Your PR was complicated not in the count of lines, but in count of commits, most of them looks like you doing try-and-run procedure. I do not forbid you to do this in your work, while you are looking for a solution or awaiting for a review comments, moreover, sometimes that is impossible without that because of the nature of CI changing moments.

I do the same things internally, but this does not mean that after addressing review comments we should keep all that crap in the main history. It has absolutely no value, hard for understanding (because that is your and only your brainstorming which is relevant only for a concrete time moment) and for maintaining.

That code is just couldn't be written in more than 100 lines because of the project codestyle.

----
In general, this claim of yours looks like a child offense, please do not do so. It only makes it worse for you. No offense.

StoneCypher replied:
> Your PR was complicated not in the count of lines, but in count of commits

I'm sorry that you find github challenging.  I gave you clear instructions on how to view it merged.  I hope someday you learn this valuable skill.

It's less than a third the size of your single function.  

&nbsp;

> most of them looks like you doing try-and-run procedure.

Each one of them describes a separate step in the process.  None of them roll back the behavior of any of the others.

I'm sorry that you find guessing and insults to be appropriate.  The reason this approach is a github standard is so that people can look at individual steps.

&nbsp;

> I do the same things internally, but this does not mean that after addressing review comments we should keep all that crap in the main history.

I'm sorry that you find Github challenging.  None of that would have gone into the main history, because of "merge on commit," which was already clearly explained to you.

I'm sorry that every time you disagree with something, or don't understand the value that it provides, you misrepresent it as crap, or the result of guesswork, to establish social dominance.

The value is clear, obvious, and quite commonly used.  I hope that some day you can learn from the people trying to teach you how to get more results out of these tools, instead of referring to their work as "crap," "low quality," and "sloppy" in public.

I'd love to see you successfully complete the same work without help, while also avoiding the choices I made that you think are incorrect.  I've been politely inviting you to for several days, and saying that perhaps I have something to learn.

&nbsp;

> It has absolutely no value

You might be surprised.

&nbsp;

> hard for understanding

You can click one tab to make it all go away, friend.  I hope you learn how some day.

&nbsp;

> because that is your and only your brainstorming

It actually isn't.  It's a valuable device for engineers to communicate about individual steps distinctly.

&nbsp;

> and for maintaining.

You keep saying there's a maintenance problem in accepting a PR that will be squashed, if it hasn't been squashed yet.

I'd be delighted to learn what, specifically that is.

StoneCypher replied:
Amusingly, what you misrepresent as a "try and run approach" is exactly the splitting up you wanted into a bunch of PRs

It's the thing you called for; you just don't realize you can have both at the same time

Mingun replied:
> It's less than a third the size of your single function.

I think it is better to discuss each PR problems in their own tickets.

> I'm sorry that you find guessing and insults to be appropriate. 

I'm sorry if you heard this in my words. I'm not a native speaker, so may be I've choose the wrong words. Forgive me.

> You can click one tab to make it all go away, friend. I hope you learn how some day.
> It actually isn't. It's a valuable device for engineers to communicate about individual steps distinctly.

It would be if each commit was followed by clear and complete description why it is done in that way and why is can not be done in the other. I've not saw that that is why I think it is your internal work, which is not necessary to be in the final history.

StoneCypher reviewed (changes requested):
Creating a good PR is not a simple task, yes. I don't think we should rush to merge unpolished PRs. 

The difference, of course, is I'm able to say specifically what isn't acceptable, and why.

Mingun reviewed (commented):
@StoneCypher, thanks for review! I'll add some comments describing the overall idea of the `generate-bytecode` pass, I hope, that saves your time for the next time

StoneCypher commented:
I just wanted you to realize how it felt to have one interaction similar to the seven days you've given me.

I think I've made my point.  Move forward with your PR as you see fit.

StoneCypher commented:
You're resolving things in bad faith

Mingun commented:
I've add some comments, explaining some details how the `generate-bytecode` pass works and fixed an error: it is impossible to remove always failed alternatives that lead to call of the custom code because it may have side effects. Added tests that proves this (`run side-effects in non-matching` behavior tests).

Ready for final review.

StoneCypher commented:
As my comment that's incorrectly marked "off-topic" says, you're trying to merge things with known bugs, in bad faith

hildjj reviewed (commented):
These are all minor comments.  One substantive point of discussion, though: shouldn't many of these cases be warnings (or errors if you pass in a "all warnings are errors" flag)?  We don't currently have a mechanism for warnings, but if we're going to do multiple errors in #160, could we discuss also adding warnings at the same time?

Mingun commented on lib/compiler/passes/inference-match-result.js:
```diff
@@ -0,0 +1,110 @@
+
+      return node.match;
+    },
+    named:        inferenceExpression,
```
I think it is better to retain the same order as in the `visitor.js` for defaults because then it is easy to visually check that all necessary methods is implemented

Mingun commented on lib/compiler/passes/inference-match-result.js:
```diff
@@ -0,0 +1,110 @@
+          // state to another.
+          // After 6 iterations the cycle with guarantee begins
+          // istanbul ignore next  This is canary test, shouldn't trigger in real life
+          if (++count > 6) {
```
Actually, I'm not sure this condition is reachable.  Proof of this goes beyond my qualifications, probably 😄 . So I've just added a canary test in case that this can occured.

Mingun commented on lib/compiler/passes/inference-match-result.js:
```diff
@@ -0,0 +1,110 @@
+              node.location
+            );
+          }
+        } while (oldResult !== node.match);
```
When you calculate the result for a recursive rule, for example;
```peggy
start = [] start
```
On the first enter `node[name="start"].match` is `undefined`, so we set it to `0` and run the detection:
https://github.com/peggyjs/peggy/blob/36e9b07a3eab37f3342efa48f2b2a9a1dc97c7cb/lib/compiler/passes/inference-match-result.js#L49-L54

Because the rule is recursive it uses it own `match` property that we speculatively set to `0`, which may not be true (as in the example where the rule contains never matched expression so should have `match=-1`). So, when we calculate our new result we start recalculation, what will use our new result, which could lead to that we calculate the third result and so on. So, for example, our assessment about `node.match` could, for example, changing in this manner:
0 -> 1 -> 0 -> -1 -> 0 -> 1 (cycle)

I this simple example we only changed from `0` to `-1` and that result never changed again. But in more complex cases that includes several recursive rules this may not be the case.

hildjj replied:
Please include a comment like:

```js
// an input of `start = [] start` will generate the sequence: 0 -> 1 -> 0 -> -1 -> 0 -> 1 (then cycle)
```

and I think we'll be able to remember what's going on when we see this again in a few years. :)

Mingun replied:
That is only an example, `start = [] start` will generate sequence `0 -> -1` :smile: But I understand you intentions. You can add this yourself and merge if this is all that needed. I can do that only tomorrow

Mingun commented:
Locally fixed all except constants. Will do that tomorrow.

Mingun commented on lib/compiler/passes/inference-match-result.js:
```diff
@@ -0,0 +1,110 @@
+const GrammarError = require("../../grammar-error");
+
+// Inference match result of the each node. Can be:
+// -1: negative result, matching of that node always fails
```
Switched to consts

Mingun commented on lib/compiler/passes/inference-match-result.js:
```diff
@@ -0,0 +1,110 @@
+
+      // If property not yet calculated, do that
+      if (typeof node.match === "undefined") {
+        node.match = 0;
```
Added

Mingun commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -248,7 +269,10 @@ function generateBytecode(ast) {
 
-  function buildCondition(condCode, thenCode, elseCode) {
+  function buildCondition(match, condCode, thenCode, elseCode) {
+    if (match > 0) { return thenCode; }
```
Done

hildjj commented on lib/compiler/passes/generate-bytecode.js:
```diff
@@ -248,7 +270,10 @@ function generateBytecode(ast) {
 
-  function buildCondition(condCode, thenCode, elseCode) {
+  function buildCondition(match, condCode, thenCode, elseCode) {
+    if (match === ALWAYS_MATCH) { return thenCode; }
```
This reads a lot easier.  thanks for that.

hildjj commented:
I went to add type tests for MatchResult, and found an error in the compile interface of peg.d.ts.  Here's an appended patch that fixes that, and adds the comment I was talking about.  Let me know your opinion, please.

Mingun commented:
Ok. I've just split you commit into two and fix a small incorrectness in you understanding of a situation about possible cycle
